### ISSUE TITLE
Use before assignment checking fails for some nested scopes

### ISSUE THREAD

**Issue by JakeStanger** [microsoft/TypeScript#42036]
**TypeScript Version:**  3.9.7 - 4.2.0-dev.20201211

**Search Terms:** used before assigned, before its declaratrion

**Code**

```ts
const someArray: any[] = [];

someArray.forEach(() => {
    someOtherArray.push('this should not work'); // does not error
});

someOtherArray.push('this should not work'); // errors as expected

const someOtherArray = [];
```

**Expected behavior:**

Both `someOtherArray.push` calls should cause a compilation error, since both occur before the variable is declared.

**Actual behavior:**

The push call just above the declaration causes a compilation as expected:
```
const someOtherArray: any[]
Block-scoped variable 'someOtherArray' used before its declaration.(2448)
input.tsx(9, 7): 'someOtherArray' is declared here.
Variable 'someOtherArray' is used before being assigned.(2454)
```

The call inside the `.forEach` compiles without a problem, but of course errors at runtime.

The exact conditions required to make it fail are unclear. Some testing has shown that `.forEach`, `.map`, `.reduce` and seemingly any method which is a part of the `someArray` prototype seems to cause the error to be ignored, while simply creating a block or IIFEs do not:

```ts
(() => {
  someOtherArray.push('this should not work'); // errors as expected
})()

const someOtherArray = [];
```
```ts
{
  someOtherArray.push('this should not work'); // errors as expected
}

const someOtherArray = [];
```

**Playground Link:** [Link](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201211#code/MYewdgzgLgBBIFsCmBBATmghgTwFw0zGwG0BdGAXhjIG4AoO+ZdLbAOgDMQ0BRTYABYAKIQEpKAPhgBvOjHlxESAPJQBSNCxxsADgFcIwgORqAlhDgCQegDYATGGBCwA7twDWR0TRgB6XzB2IEgWTrAaaNx0AL7eDCLiFFKy8kwqahpa7PqGQiYC5pbW9o7OMG5ont5+ARHcFpgWSAAeOkjAUEh2MaJi9HQpisiq6poY2jnGZhaGxQ5h5R5ePv4wdWgNTa3tnd3RDGkjmePZBlMFM1a282UVVSu1GPUEW20dXQygkLCHGWOslGopBoQA)

**Comment by MartinJohns** [comment]
#9998

**Comment by RyanCavanaugh** [comment]
See #11498; we don't error here because it *might* be valid and there'd be no way to indicate when this was the case if we errored here.
